File: leetcode/medium/9999_testing_script/MySol.swift
import Foundation

class Solution {
    func findKthLargest(_ nums: [Int], _ k: Int) -> Int {
        var minHeap = Heap(priorityFunction: >, elements: Array(nums[0..<k]))
        
        for i in k..<nums.count {
            if nums[i] > minHeap.peek()! {
                minHeap.replace(element: nums[i])
            }
        }
        
        return minHeap.peek()!
    }
}

struct Heap<Element> {
    var elements: [Element]
    let priorityFunction: (Element, Element) -> Bool
    
    init(priorityFunction: @escaping (Element, Element) -> Bool, elements: [Element] = []) {
        self.priorityFunction = priorityFunction
        self.elements = elements
        buildHeap()
    }
    
    var isEmpty: Bool {
        return elements.isEmpty
    }
    
    var count: Int {
        return elements.count
    }
    
    func peek() -> Element? {
        return elements.first
    }
    
    mutating func buildHeap() {
        for i in stride(from: (elements.count / 2 - 1), through: 0, by: -1) {
            siftDown(from: i)
        }
    }
    
    mutating func siftDown(from index: Int) {
        var parent = index
        while true {
            let left = 2 * parent + 1
            let right = 2 * parent + 2
            var candidate = parent
            if left < elements.count && priorityFunction(elements[left], elements[candidate]) {
                candidate = left
            }
            if right < elements.count && priorityFunction(elements[right], elements[candidate]) {
                candidate = right
            }
            if candidate == parent {
                return
            }
            elements.swapAt(parent, candidate)
            parent = candidate
        }
    }
    
    mutating func replace(element: Element) {
        elements[0] = element
        siftDown(from: 0)
    }
}

File: leetcode/medium/9999_testing_script/MySol.rb
class Solution
  def find_kth_largest(nums, k)
    min_heap = MinHeap.new(nums[0...k])
    
    (k...nums.length).each do |i|
      min_heap.push(nums[i]) if nums[i] > min_heap.peek
    end
    
    min_heap.peek
  end
end

class MinHeap
  attr_reader :heap
  
  def initialize(array = [])
    @heap = build_heap(array)
  end
  
  def build_heap(array)
    last_parent = (array.length - 2) / 2
    (last_parent).downto(0).each do |i|
      sift_down(i, array.length - 1, array)
    end
    array
  end
  
  def sift_down(start, end_idx, heap)
    root = start
    while root * 2 + 1 <= end_idx
      child = root * 2 + 1
      swap = root
      swap = child if heap[swap] > heap[child]
      swap = child + 1 if child + 1 <= end_idx && heap[swap] > heap[child + 1]
      break if swap == root
      heap[root], heap[swap] = heap[swap], heap[root]
      root = swap
    end
  end
  
  def peek
    heap.first
  end
  
  def push(value)
    heap.push(value)
    sift_down(0, heap.length - 1, heap)
  end
end
